service/state/rpc.go

```diff
@@ -80,24 +100,77 @@ func (s *Service) handleSubmitTx(w http.ResponseWriter, r *http.Request) {
 	raw, err := hex.DecodeString(txStr)
 	if err != nil {
 		w.WriteHeader(http.StatusBadRequest)
-		log.Errorw("serving /submit_tx request", "err", err)
+		log.Errorw("serving request", "endpoint", submitTxEndpoint, "err", err)
 		return
 	}
 	// perform request
 	txResp, err := s.accessor.SubmitTx(r.Context(), raw)
 	if err != nil {
 		w.WriteHeader(http.StatusInternalServerError)
-		log.Errorw("serving /submit_tx request", "err", err)
+		w.Write([]byte(err.Error())) //nolint:errcheck
+		log.Errorw("serving request", "endpoint", submitTxEndpoint, "err", err)
+		return
+	}
+	resp, err := json.Marshal(txResp)
+	if err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		log.Errorw("serving request", "endpoint", submitTxEndpoint, "err", err)
+		return
+	}
+	_, err = w.Write(resp)
+	if err != nil {
+		log.Errorw("writing response", "endpoint", submitTxEndpoint, "err", err)
+	}
+}
+
+func (s *Service) handleSubmitPFD(w http.ResponseWriter, r *http.Request) {
+	// parse body from request
+	raw, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		w.WriteHeader(http.StatusBadRequest)
+		log.Errorw("serving request", "endpoint", submitPFDEndpoint, "err", err)
+		return
+	}
+	// decode request
+	req := new(submitPFDRequest)
+	err = json.Unmarshal(raw, req)
+	if err != nil {
+		w.WriteHeader(http.StatusBadRequest)
+		log.Errorw("serving request", "endpoint", submitPFDEndpoint, "err", err)
+		return
+	}
+
+	nID, err := hex.DecodeString(req.NamespaceID)
+	if err != nil {
+		w.WriteHeader(http.StatusBadRequest)
+		log.Errorw("serving request", "endpoint", submitPFDEndpoint, "err", err)
+		return
+	}
+	data, err := hex.DecodeString(req.Data)
+	if err != nil {
+		w.WriteHeader(http.StatusBadRequest)
+		log.Errorw("serving request", "endpoint", submitPFDEndpoint, "err", err)
+		return
+	}
+	// perform request
+	txResp, err := s.accessor.SubmitPayForData(r.Context(), nID, data, req.GasLimit)
+	if err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		_, werr := w.Write([]byte(err.Error()))
+		if werr != nil {
+			log.Errorw("writing response", "endpoint", submitPFDEndpoint, "err", werr)
+		}
```

Comment:
For the sake of consistency - //nolint:errcheck or do actual err check in the error writes. Ideally add errchecks everywhere, so that if something very wrong happens, we could debug it from logs

Reply:
done